File: solutions/_129_sum_root_to_leaf_numbers/solution.py
from typing import Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        return self.dfs(root, 0)

    def dfs(self, node, total):
        if not node:
            return 0
        total = total * 10 + node.val
        if not node.left and not node.right:
            return total
        return self.dfs(node.left, total) + self.dfs(node.right, total)


# https://leetcode.com/problems/sum-root-to-leaf-numbers/

File: solutions/_129_sum_root_to_leaf_numbers/Solution.java

package _129_sum_root_to_leaf_numbers;

class Solution {
    public int sumNumbers(TreeNode root) {
        return dfs(root, 0);
    }

    private int dfs(TreeNode node, int sum) {
        if (node == null) {
            return 0;
        }
        sum = sum * 10 + node.val;
        if (node.right == null && node.left == null) {
            return sum;
        }
        return dfs(node.left, sum) + dfs(node.right, sum);
    }
}

/*
 * https://leetcode.com/problems/sum-root-to-leaf-numbers/
 */
